addon/classes/ChangesetArrayProxyHandler.ts

```diff
@@ -0,0 +1,298 @@
+      case 'at': return this.at;
+      case 'concat': return this.readArray.concat;
+      case 'length': return this.readArray.length;
+      case 'copyWithin': return this.writeArray.copyWithin;
+      case 'entries': return this.readArray.entries;
+      case 'every': return this.readArray.every;
+      case 'fill': return this.writeArray.fill;
+      case 'filter': return this.readArray.filter;
+      case 'find': return this.readArray.find;
+      case 'findIndex': return this.readArray.findIndex;
+      case 'flat': return this.readArray.flat;
+      case 'flatMap': return this.readArray.flatMap;
+      case 'forEach': return this.readArray.forEach;
+      case 'includes': return this.readArray.includes;
+      case 'indexOf': return this.readArray.indexOf;
+      case 'join': return this.readArray.join;
+      case 'keys': return this.readArray.keys;
+      case 'lastIndexOf': return this.readArray.lastIndexOf;
+      case 'map': return this.readArray.map;
+      case 'pop': return this.writeArray.pop;
+      case 'push': return this.writeArray.push;
+      case 'reduce': return this.readArray.reduce;
+      case 'reduceRight': return this.readArray.reduceRight;
+      case 'reverse': return this.writeArray.reverse;
+      case 'shift': return this.writeArray.shift;
+      case 'slice': return this.readArray.slice;
+      case 'some': return this.readArray.some;
+      case 'sort': return this.writeArray.sort;
+      case 'splice': return this.writeArray.splice;
+      case 'toLocaleString': return this.readArray.toLocaleString;
+      case 'toString': return this.readArray.toString;
+      case 'unshift': return this.writeArray.unshift;
+      case 'values': return this.readArray.values;
+
+      case 'change': return this.change;
+      case 'changes': return this.changes;
+      case 'data': return this.__source;
+      case 'execute': return this.execute;
+      case 'get': return this.getValue;
+      case 'isChangeset': return true;
+      case 'isDirty': return this.isDirty;
+      case 'isPristine': return this.isPristine;
+      case 'isValid': return this.isValid;
+      case 'pendingData': return this.pendingData;
+      case 'rollback': return this.rollback;
+      case 'save': return this.save;
+      case 'set': return this.setValue;
+      case 'validate': return this.validate;
+      default: return this.getValue(key);
+    }
+  }
+
+  public set(_target: T, key: string, value: any): any {
+    return this.setValue(key, value);
+  }
+
+  public readonly isChangeset = true;
+
+  public get isDirty(): boolean {
+    // we're dirty if either we have top level changes
+    // or if a nested proxy is dirty
+    return (this.__proxyArray !== undefined);
+  }
+
+  public get isPristine(): boolean {
+    return (this.__proxyArray === undefined);
+  }
+
+  public get isValid(): boolean {
+    // TODO: validate this level
+
+    // now look at all the nested proxies
+    for (var proxy of this.__nestedProxies.values()) {
```

Comment:
Why reimplement these behaviors? Couldn't you return tte value from the underlying validated-changeset?